Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -918,6 +966,146 @@ lemma nonsingular [Nontrivial R] {x y : R} (h : E.equation x y) : E.nonsingular
+  ⟨
+    WeierstrassCurve.ofJInvariant0 R,
+    ⟨-27, -(⅟3) ^ 3,
+      by
+        rw [calc (27 : R) = (3 : R) ^ 3 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 3 (mul_invOf_self (3 : R))
+      , by
+        rw [calc (27 : R) = (3 : R) ^ 3 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 3 (invOf_mul_self (3 : R))
+    ⟩,
+    by rw [WeierstrassCurve.ofJInvariant0_Δ R]
+  ⟩
+
+lemma ofJInvariant0_j [Invertible (3 : R)] : (ofJInvariant0 R).j = 0 := by
+  simp only [j, ofJInvariant0, WeierstrassCurve.ofJInvariant0_c₄]
+  ring1
+
+/-- The elliptic curve $Y^2 = X^3 + X$ when $2$ is invertible. -/
+def ofJInvariant1728 [Invertible (2 : R)] : EllipticCurve R :=
+  ⟨
+    WeierstrassCurve.ofJInvariant1728 R,
+    ⟨-64, -(⅟2) ^ 6,
+      by
+        rw [calc (64 : R) = (2 : R) ^ 6 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 6 (mul_invOf_self (2 : R))
+      , by
+        rw [calc (64 : R) = (2 : R) ^ 6 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 6 (invOf_mul_self (2 : R))
+    ⟩,
+    by rw [WeierstrassCurve.ofJInvariant1728_Δ R]
+  ⟩
+
+lemma ofJInvariant1728_j [Invertible (2 : R)] : (ofJInvariant1728 R).j = 1728 := by
+  simp only [j, ofJInvariant1728, WeierstrassCurve.ofJInvariant1728_c₄]
+  field_simp
+  ring1
+
+end ModelsWithJInvariant_0_1728
+
+section ModelsWithJInvariant'
+
+variable {R : Type u} [CommRing R] (j : R) [Invertible j] [Invertible (j - 1728)]
+
+/-- The elliptic curve $Y^2 + (j-1728)XY = X^3 - 36(j-1728)^3X - (j-1728)^5$
+when $j$ and $j-1728$ are both invertible. -/
+def ofJInvariant' : EllipticCurve R :=
+  ⟨
+    WeierstrassCurve.ofJInvariant j,
+    ⟨j ^ 2 * (j - 1728) ^ 9, ⅟j ^ 2 * ⅟(j - 1728) ^ 9,
+      by rw [←mul_assoc, mul_assoc (j ^ 2) ((j - 1728) ^ 9) (⅟j ^ 2),
+        mul_comm ((j - 1728) ^ 9) (⅟j ^ 2), ←mul_assoc, mul_assoc,
+        pow_mul_pow_eq_one 2 (mul_invOf_self _),
+        pow_mul_pow_eq_one 9 (mul_invOf_self _), mul_one]
+      , by rw [←mul_assoc, mul_assoc (⅟j ^ 2) (⅟(j - 1728) ^ 9) (j ^ 2),
+        mul_comm (⅟(j - 1728) ^ 9) (j ^ 2), ←mul_assoc, mul_assoc,
+        pow_mul_pow_eq_one 2 (invOf_mul_self _),
+        pow_mul_pow_eq_one 9 (invOf_mul_self _), mul_one]
+    ⟩,
+    by rw [WeierstrassCurve.ofJInvariant_Δ j]
+  ⟩
+
+lemma ofJInvariant'_j : (ofJInvariant' j).j = j := by
+  simp only [EllipticCurve.j, ofJInvariant', WeierstrassCurve.ofJInvariant_c₄]
+  field_simp
+  -- do not use 'ring1' otherwise (j - 1728) ^ 9 will be expanded
+  rw [mul_pow, ←pow_mul, ←mul_assoc, ←pow_succ]
+
+end ModelsWithJInvariant'
+
+section ModelsWithJInvariant
+
+variable {F : Type u} [Field F] (j : F)
+
+/-- For any element $j$ of a field $F$, there exists an elliptic curve over $F$
+with $j$-invariant equal to $j$. Its coefficients are given explicitly. -/
+noncomputable def ofJInvariant : EllipticCurve F := by
```

Comment:
I think given this is a `def`, it's probably wise not to have it in tactic mode, although it doesn't seem like your proof has any dangerous tactics (e.g. `rw` - in your case is only used to prove things are invertible, which is fine). I'll think of how to make this into term mode when I have time.

Reply:
In fact in my previous attempt there is no such `def`, but only a theorem which states that `... (j : F) : \exists E : EllipticCurve, E.j = j`. A disadvantage is that you can't get more information of such `E`, compared to current implementation. This is what `coefficients are given explicitly` means.

> I'll think of how to make this into term mode when I have time.

Thanks.